src/scene/materials/shader-graph-builder.js

```diff
@@ -0,0 +1,128 @@
+import { ShaderGraphNode } from './shader-graph-node.js';
+
+var id = 0;
+
+/**
+ * @private
+ * @class
+ * @name pc.ShaderGraphBuilder
+ * @classdesc A Shader Graph Builder class
+ * @param {pc.Application} app - Application with shader graph core nodes registered that will be used to build the shader graph
+ */
+var ShaderGraphBuilder = function (app) {
+    id++;
+
+    this.app = app;
+    this._nodes = this.app.shaderNodes._nodes;
+
+    this._graph = new ShaderGraphNode();
+    this._graph.name = 'graphRoot_' + id;
+};
+
+ShaderGraphBuilder.prototype.constructor = ShaderGraphBuilder;
+
+Object.assign(ShaderGraphBuilder.prototype, {
+    /**
+     * @private
+     * @function
+     * @name pc.ShaderGraphBuilder#addParam
+     * @description adds a parameter input to graph
+     * @param {string} type - type of parameter
+     * @param {string} name - name of parameter
+     * @param {any} value - optional value of parameter
+     * @returns {any} returns the created input port
+     */
+    addParam: function (type, name, value) {
+        var ioPort = this._graph.addInput(type, name, value);
+        return ioPort;
+    },
+    /**
+     * @private
+     * @function
+     * @name pc.ShaderGraphBuilder#addNode
+     * @description creates and adds a core node to the shader graph and connects up inputs
+     * @param {string} name - name of parameter
+     * @param {object} [arguments] - first argument is name string, then 0-N output node ports or params to connect to new core node inputs - number and port type of arguments must match.
+     * @returns {number} returns the created node id
+     * @example
+     * var node0 = pc.shadergraph.addNode('uv0');
+     * var node1 = pc.shadergraph.addNode('texSample', param0, node0 );
+     * var node2 = pc.shadergraph.addNode('splitVec4', node1 );
+     * var node3 = pc.shadergraph.addNode('mul3', param1, node42 );
+     */
+    addNode: function (name) {
+        var args = arguments;
+        var coreNode = this._nodes[name];
+        var sgIndex = this._graph.addSubGraph(coreNode);

```

Comment:
I think the public API should be more robust and check / report issues. In this case coreNode can be null, which would cause exception later. addSubGraph should handle the case of duplicate and print error / fail and similar, to help users diagnose issues.

Reply:
A check would be nice - when/if the API is public a badly typed in name could result in exceptions - what could be better is a enum of core nodes (which isn't possible yet because the complete list of core nodes hasn't yet been locked down) - for now - when used privately with the editor - its less of a problem?

Reply:
Validation has been added